lib/core/src/Cardano/Pool/Ranking.hs

```diff
@@ -0,0 +1,206 @@
+-- +------------+-------------+------------------------------------------------+
+-- | f          | @ [0, ∞) @  | rewards                                        |
+-- +------------+-------------+------------------------------------------------+
+-- | m          | @ [0, 1] @  | margin                                         |
+-- +------------+-------------+------------------------------------------------+
+-- | p_apparent | @ [0, 1] @  | apparent performance                           |
+-- +------------+-------------+------------------------------------------------+
+-- | s          | @ [0, ∞) @  | relative stake of the pool leader (i.e pledge) |
+-- +------------+-------------+------------------------------------------------+
+-- | σ          | @ [0, 1] @  | total relative stake of the pool               |
+-- +------------+-------------+------------------------------------------------+
+module Cardano.Pool.Ranking
+    (
+      -- * Formulas
+      desirability
+
+    , saturatedPoolRewards
+    , saturatedPoolSize
+
+      -- * Types
+    , EpochConstants (..)
+    , Pool (..)
+    , RelativeStakeOf (..)
+    , mkRelativeStake
+    , Lovelace (..)
+    , Margin
+    , unsafeMkMargin
+    , getMargin
+    , NonNegative (..)
+    , Positive (..)
+    , unsafeToPositive
+    , unsafeToNonNegative
+    )
+    where
+
+import Prelude
+
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( Symbol )
+
+--------------------------------------------------------------------------------
+-- Formulas from spec
+--------------------------------------------------------------------------------
+
+-- | Non-myopic pool desirability according to section 5.6.1.
+--
+-- Is /not/ affected by oversaturation nor pool stake in general.
+desirability
+    :: EpochConstants
+    -> Pool
+    -> Double
+desirability constants pool
+    | f_saturated <= c = 0
+    | otherwise    = (f_saturated - c) * (1 - m)
+  where
+    f_saturated = saturatedPoolRewards constants pool
+    m = getMargin $ margin pool
+    c = getNonNegative $ getLovelace $ cost pool
+
+-- | Total rewards for a pool if it were saturated.
+--
+-- When a0 = 0 this reduces to just p*R*z0 (tested
+-- by @prop_saturatedPoolRewardsReduces@)
+saturatedPoolRewards :: EpochConstants -> Pool -> Double
+saturatedPoolRewards constants pool =
+    let
+        a0 = getNonNegative $ leaderStakeInfluence constants
+        z0 = unRelativeStake $ saturatedPoolSize constants
+        s = unRelativeStake $ leaderStake pool
+        _R = getNonNegative $ getLovelace $ totalRewards constants
+        p = getNonNegative $ recentAvgPerformance pool
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
+
+-- | Determines z0, i.e 1 / k
+saturatedPoolSize :: EpochConstants -> RelativeStakeOf "pool"
+saturatedPoolSize constants =
+    RelativeStake $ 1 / fromIntegral (getPositive $ optimalNumberOfPools constants)
```

Comment:
Why is this of type `RelativeStakeOf` ? I thought the point of these newtypes was to provide insights on the semantic of the various numbers, `σ` stands for the relative stake of the pool and, although they have the same dimension, labelling both as "RelativeStakeOf" feels misleading. 

Aype that captures what the domain actually is would be preferable (which is what types for in the first place, representing a range of possible values): `Ratio 1 0 ` or `Percentage` ?

Reply:
I introduced `RelativeStakeOf` to protect against mixing up `\sigma` and `s`.

Although, at that time, most functions took only the pool parameters they needed. And it was easy to get the order wrong. Now both `desirability` and `saturatedPoolSize` take `Pool`, and the `RelativeStakeOf tag` tag might be less useful.